src/Servers/Kestrel/shared/KnownHeaders.cs

```diff
@@ -185,6 +188,8 @@ static KnownHeaders()
             ResponseTrailers = new[]
             {
                 "ETag",
+                "Grpc-Message",
+                "Grpc-Status"
```

Comment:
What do you think about splitting HttpRequestHeaders into Http1RequestHeaders and Http2RequestHeaders? If we include the pseudo-header fields, there are now 9 request header fields in HttpRequestHeaders that I wouldn't expect to be used with HTTP/1.x (though maybe grpc-web means the gRPC headers are also common over HTTP/1.x now).

I'm concerned that using the same HttpRequestHeaders class for HTTP/1.x and HTTP/2 and adding too many HTTP/2-specific headers will slow down HTTP/1.x unnecessarily. We already have 4 HTTP/2 pseudo-header fields in `requestPrimaryHeaders`. TryGetValueFast will check if the requested header name matches these primary headers before other header names of the same length. Any extra header fields waste memory and slow down methods like Clear, TryGetValueFast, Append, etc...

Reply:
I think that would be a good improvement. I don't know if there will be time for .NET 5.

On primary headers: the HTTP/2 pseudo-headers can probably be removed from primary headers. Pseudo-headers will be sent in HEADERS frame as static table indexes. They'll use the upcoming optimizations here: https://github.com/dotnet/aspnetcore/pull/24730

Reply:
Are these exact header names used for both HTTP/2 gRPC and gRPC-Web? If so, I guess these warrant being in a possible Http1RequestHeaders dictionary anyway.

Reply:
gRPC-Web headers are the same, and gRPC-Web is usable in HTTP/1.1 and 2. So yes it is useful for them to be in a potential Http1RequestHeaders.

The only HTTP/2 specific headers are the pseudo ones.